esda/significance.py

```diff
@@ -0,0 +1,101 @@
+import numpy as np
+from scipy import stats
+
+
+def calculate_significance(test_stat, reference_distribution, method="two-sided"):
+    """
+    Calculate a pseudo p-value from a reference distribution.
+
+    Pseudo-p values are calculated using the formula (M + 1) / (R + 1). Where R is the number of simulations and M is the number of times that the simulated value was equal to, or more extreme than the observed test statistic.
+
+    Simulated test statistics are generated through a process of conditional permutation. Conditional permutation holds fixed the value of Xi and values of neighbors are randomly sampled from X removing Xi simulating spatial randomness. This process is repeated R times to generate a reference distribution from which the pseudo-p value is calculated.
+
+    Parameters
+    ----------
+    test_stat:
+        The observed test statistic, or a vector of observed test statistics
+    reference_distribution:
+        A numpy array containing simulated test statistics as a result of conditional permutation.
+    method:
+        One of 'two-sided', 'lesser', or 'greater'. Indicates the alternative hypothesis.
+        - 'two-sided': the observed test-statistic is more-extreme than expected under the assumption of complete spatial randomness.
+        - 'lesser': the observed test-statistic is less than the expected value under the assumption of complete spatial randomness.
+        - 'greater': the observed test-statistic is greater than the exepcted value under the assumption of complete spatial randomness.
+        - 'directed': run both lesser and greater tests, then pick the smaller p-value.
+    """
+    reference_distribution = np.atleast_2d(reference_distribution)
+    n_samples, p_permutations = reference_distribution.shape
+    test_stat = np.atleast_2d(test_stat).reshape(n_samples, -1)
+    if method == "directed":
+        larger = (reference_distribution >= test_stat).sum(axis=1)
+        low_extreme = (p_permutations - larger) < larger
+        larger[low_extreme] = p_permutations - larger[low_extreme]
+        p_value = (larger + 1.0) / (p_permutations + 1.0)
+    elif method == "lesser":
+        p_value = (np.sum(reference_distribution >= test_stat, axis=1) + 1) / (
+            p_permutations + 1
+        )
+    elif method == "greater":
+        p_value = (np.sum(reference_distribution <= test_stat, axis=1) + 1) / (
```

Comment:
Should this be >= ?

Reply:
I _think_ this order is correct because we're saying the reference distribution is less than or equal to the test stat. It may make sense to flip them just for the sake reading clarity

Reply:
Yes, I think @sjsrey is right. 

The p-value for the one sided permutation test is the fraction of replicates further out in the tail. The "alternative" refers to the alternative hypothesis that generates the test stat, not the reference distribution. 

So, for "greater", we need to count the replicates *above* the test stat... if too many are above the test stat, the p value will be large and we can't reject the null hypothesis that the test stat is *not greater* than the reference distribution.